File: pkg/ui/src/api/services/iam.ts
/**
 * IAM Service API
 * REST HTTP client for IAM via Go proxy
 */

import { PROXY_BACKEND } from '@/config'
import { APIError } from '../client'

const api = PROXY_BACKEND.replace(/\/$/, '')

// Types (matching the existing type definitions)
export interface IAMUser {
  UserName: string
  UserId: string
  Arn: string
  CreateDate: string
  Path?: string
}

export interface IAMRole {
  RoleName: string
  RoleId: string
  Arn: string
  CreateDate: string
  Path?: string
}

export interface IAMPolicy {
  PolicyName: string
  PolicyId: string
  Arn: string
  Path: string
  DefaultVersionId?: string
  AttachmentCount?: number
  PermissionsBoundaryUsageCount?: number
  IsAttachable: boolean
  Description?: string
  CreateDate: string
  UpdateDate: string
}

export interface IAMGroup {
  GroupName: string
  GroupId: string
  Arn: string
  CreateDate: string
  Path?: string
}

// User operations
export async function createUser(params: {
  UserName: string
  Path?: string
  PermissionsBoundary?: string
  Tags?: Array<{ Key: string; Value: string }>
}): Promise<IAMUser> {
  const res = await fetch(`${api}/iam/users`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(params),
  })
  if (!res.ok) throw new APIError(`Create user failed`, res.status, 'iam')
  return res.json()
}

export async function getUser(UserName?: string): Promise<IAMUser> {
  if (!UserName) throw new APIError('UserName is required for getUser', 400, 'iam')
  const res = await fetch(`${api}/iam/users/${encodeURIComponent(UserName)}`)
  if (!res.ok) throw new APIError(`Get user failed`, res.status, 'iam')
  return res.json()
}

export async function listUsers(): Promise<{ Users: IAMUser[]; IsTruncated: boolean; Marker?: string }> {
  const res = await fetch(`${api}/iam/users`)
  if (!res.ok) throw new APIError(`List users failed`, res.status, 'iam')
  const data = await res.json()
  return {
    Users: data.Users || [],
    IsTruncated: data.IsTruncated || false,
    Marker: data.Marker,
  }
}

export async function deleteUser(UserName: string): Promise<void> {
  const res = await fetch(`${api}/iam/users/${encodeURIComponent(UserName)}`, { method: 'DELETE' })
  if (!res.ok) throw new APIError(`Delete user failed`, res.status, 'iam')
}

// Role operations
export async function createRole(params: {
  RoleName: string
  AssumeRolePolicyDocument: string
  Description?: string
  MaxSessionDuration?: number
  PermissionsBoundary?: string
  Tags?: Array<{ Key: string; Value: string }>
}): Promise<IAMRole> {
  const res = await fetch(`${api}/iam/roles`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(params),
  })
  if (!res.ok) throw new APIError(`Create role failed`, res.status, 'iam')
  return res.json()
}

export async function getRole(RoleName: string): Promise<IAMRole> {
  const res = await fetch(`${api}/iam/roles/${encodeURIComponent(RoleName)}`)
  if (!res.ok) throw new APIError(`Get role failed`, res.status, 'iam')
  return res.json()
}

export async function listRoles(): Promise<{ Roles: IAMRole[]; IsTruncated: boolean; Marker?: string }> {
  const res = await fetch(`${api}/iam/roles`)
  if (!res.ok) throw new APIError(`List roles failed`, res.status, 'iam')
  const data = await res.json()
  return {
    Roles: data.Roles || [],
    IsTruncated: data.IsTruncated || false,
    Marker: data.Marker,
  }
}

export async function deleteRole(RoleName: string): Promise<void> {
  const res = await fetch(`${api}/iam/roles/${encodeURIComponent(RoleName)}`, { method: 'DELETE' })
  if (!res.ok) throw new APIError(`Delete role failed`, res.status, 'iam')
}

// Policy operations
export async function listPolicies(param1?: string | { Scope?: string; OnlyAttached?: boolean }, param2?: boolean): Promise<{
  Policies: IAMPolicy[]
  IsTruncated: boolean
  Marker?: string
}> {
  const params = new URLSearchParams()

  // Handle both calling conventions: listPolicies('All') or listPolicies({ Scope: 'All' })
  if (typeof param1 === 'string') {
    params.set('Scope', param1)
    if (param2 !== undefined) params.set('OnlyAttached', String(param2))
  } else if (param1 && typeof param1 === 'object') {
    if (param1.Scope) params.set('Scope', param1.Scope)
    if (param1.OnlyAttached !== undefined) params.set('OnlyAttached', String(param1.OnlyAttached))
  }

  const qs = params.toString()
  const url = qs ? `${api}/iam/policies?${qs}` : `${api}/iam/policies`
  const res = await fetch(url)
  if (!res.ok) throw new APIError(`List policies failed`, res.status, 'iam')

  const data = await res.json()
  const policies = (data.Policies || data.policies || []).map((p: any) => ({
    ...p,
    PolicyArn: p.Arn || p.PolicyArn,
  }))
  return {
    Policies: policies,
    IsTruncated: data.IsTruncated || data.isTruncated || false,
    Marker: data.Marker || data.marker,
  }
}

export async function getPolicy(PolicyArn: string): Promise<any> {
  const res = await fetch(`${api}/iam/policies/get`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ PolicyArn }),
  })
  if (!res.ok) throw new APIError(`Get policy failed`, res.status, 'iam')
  return res.json()
}

export async function createPolicy(input: CreatePolicyInput): Promise<any> {
  const res = await fetch(`${api}/iam/policies`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(input),
  })
  if (!res.ok) throw new APIError(`Create policy failed`, res.status, 'iam')
  return res.json()
}

export async function deletePolicy(PolicyArn: string): Promise<void> {
  const res = await fetch(`${api}/iam/policies/delete`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ PolicyArn }),
  })
  if (!res.ok) throw new APIError(`Delete policy failed`, res.status, 'iam')
}

// Access Key operations
export async function createAccessKey(UserName: string): Promise<any> {
  const res = await fetch(`${api}/iam/access-keys`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ UserName }),
  })
  if (!res.ok) throw new APIError(`Create access key failed`, res.status, 'iam')
  return res.json()
}

export async function listAccessKeys(UserName?: string): Promise<any> {
  const url = UserName
    ? `${api}/iam/access-keys?UserName=${encodeURIComponent(UserName)}`
    : `${api}/iam/access-keys`
  const res = await fetch(url)
  if (!res.ok) throw new APIError(`List access keys failed`, res.status, 'iam')
  return res.json()
}

export async function deleteAccessKey(AccessKeyId: string, UserName?: string): Promise<void> {
  const res = await fetch(`${api}/iam/access-keys/${encodeURIComponent(AccessKeyId)}`, { method: 'DELETE' })
  if (!res.ok) throw new APIError(`Delete access key failed`, res.status, 'iam')
}

export async function updateAccessKeyStatus(AccessKeyId: string, Status: 'Active' | 'Inactive', UserName?: string): Promise<void> {
  const res = await fetch(`${api}/iam/access-keys/${encodeURIComponent(AccessKeyId)}`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ Status, UserName }),
  })
  if (!res.ok) throw new APIError(`Update access key status failed`, res.status, 'iam')
}

// Role policy operations
export async function attachRolePolicy(RoleName: string, PolicyArn: string): Promise<void> {
  const res = await fetch(`${api}/iam/roles/${encodeURIComponent(RoleName)}/policies`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ PolicyArn }),
  })
  if (!res.ok) throw new APIError(`Attach role policy failed`, res.status, 'iam')
}

export async function detachRolePolicy(RoleName: string, PolicyArn: string): Promise<void> {
  const res = await fetch(`${api}/iam/roles/${encodeURIComponent(RoleName)}/detach-policy`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ PolicyArn }),
  })
  if (!res.ok) throw new APIError(`Detach role policy failed`, res.status, 'iam')
}

export async function listAttachedRolePolicies(RoleName: string): Promise<any> {
  const res = await fetch(`${api}/iam/roles/${encodeURIComponent(RoleName)}/policies`)
  if (!res.ok) throw new APIError(`List attached role policies failed`, res.status, 'iam')
  return res.json()
}

// Group operations
export async function createGroup(GroupName: string, Path?: string): Promise<IAMGroup> {
  const res = await fetch(`${api}/iam/groups`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ GroupName, Path }),
  })
  if (!res.ok) throw new APIError(`Create group failed`, res.status, 'iam')
  return res.json()
}

export async function getGroup(GroupName: string): Promise<{ Group: IAMGroup; Users?: IAMUser[]; IsTruncated: boolean }> {
  const res = await fetch(`${api}/iam/groups/${encodeURIComponent(GroupName)}`)
  if (!res.ok) throw new APIError(`Get group failed`, res.status, 'iam')
  return res.json()
}

export async function listGroups(): Promise<{ Groups: IAMGroup[]; IsTruncated: boolean }> {
  const res = await fetch(`${api}/iam/groups`)
  if (!res.ok) throw new APIError(`List groups failed`, res.status, 'iam')
  const data = await res.json()
  return {
    Groups: data.Groups || [],
    IsTruncated: data.IsTruncated || false,
  }
}

export async function deleteGroup(GroupName: string): Promise<void> {
  const res = await fetch(`${api}/iam/groups/${encodeURIComponent(GroupName)}`, { method: 'DELETE' })
  if (!res.ok) throw new APIError(`Delete group failed`, res.status, 'iam')
}

export async function addUserToGroup(GroupName: string, UserName: string): Promise<void> {
  const res = await fetch(`${api}/iam/groups/${encodeURIComponent(GroupName)}/users`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ UserName }),
  })
  if (!res.ok) throw new APIError(`Add user to group failed`, res.status, 'iam')
}

export async function removeUserFromGroup(GroupName: string, UserName: string): Promise<void> {
  const res = await fetch(`${api}/iam/groups/${encodeURIComponent(GroupName)}/users/${encodeURIComponent(UserName)}`, { method: 'DELETE' })
  if (!res.ok) throw new APIError(`Remove user from group failed`, res.status, 'iam')
}

export async function listGroupsForUser(UserName: string): Promise<{ Groups: IAMGroup[]; IsTruncated: boolean }> {
  const res = await fetch(`${api}/iam/users/${encodeURIComponent(UserName)}/groups`)
  if (!res.ok) throw new APIError(`List groups for user failed`, res.status, 'iam')
  const data = await res.json()
  return {
    Groups: data.Groups || [],
    IsTruncated: data.IsTruncated || false,
  }
}

export async function listUsersForGroup(GroupName: string): Promise<{ Users: any[]; IsTruncated: boolean }> {
  const res = await fetch(`${api}/iam/groups/${encodeURIComponent(GroupName)}/users`)
  if (!res.ok) throw new APIError(`List users for group failed`, res.status, 'iam')
  const data = await res.json()
  return {
    Users: data.Users || [],
    IsTruncated: data.IsTruncated || false,
  }
}

// Inline policy operations
export async function listUserPolicies(UserName: string): Promise<any> {
  const res = await fetch(`${api}/iam/users/${encodeURIComponent(UserName)}/policies`)
  if (!res.ok) throw new APIError(`List user policies failed`, res.status, 'iam')
  return res.json()
}

export async function listRolePolicies(RoleName: string): Promise<any> {
  const res = await fetch(`${api}/iam/roles/${encodeURIComponent(RoleName)}/inline-policies`)
  if (!res.ok) throw new APIError(`List role policies failed`, res.status, 'iam')
  return res.json()
}

export async function getRolePolicy(RoleName: string, PolicyName: string): Promise<any> {
  const res = await fetch(`${api}/iam/roles/${encodeURIComponent(RoleName)}/policies/${encodeURIComponent(PolicyName)}`)
  if (!res.ok) throw new APIError(`Get role policy failed`, res.status, 'iam')
  return res.json()
}

export interface CreatePolicyInput {
  PolicyName: string
  PolicyDocument: string
  Description?: string
}

File: pkg/proxy/internal/adapters/http/iam.go
package httphandlers

import (
	"net/http"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/iam"
	iamtypes "github.com/aws/aws-sdk-go-v2/service/iam/types"
	"github.com/go-chi/chi/v5"
)

func (h *ProxyHandler) registerIAMRoutes(r chi.Router) {
	r.Route("/iam", func(r chi.Router) {
		r.Get("/users", h.listUsers)
		r.Post("/users", h.createUser)
		r.Get("/users/{userName}", h.getUser)
		r.Delete("/users/{userName}", h.deleteUser)

		r.Post("/access-keys", h.createAccessKey)
		r.Get("/access-keys", h.listAccessKeys)
		r.Delete("/access-keys/{accessKeyId}", h.deleteAccessKey)
		r.Put("/access-keys/{accessKeyId}", h.updateAccessKeyStatus)

		r.Get("/roles", h.listRoles)
		r.Post("/roles", h.createRole)
		r.Get("/roles/{roleName}", h.getRole)
		r.Delete("/roles/{roleName}", h.deleteRole)

		r.Post("/roles/{roleName}/policies", h.attachRolePolicy)
		r.Post("/roles/{roleName}/detach-policy", h.detachRolePolicy)
		r.Get("/roles/{roleName}/policies", h.listAttachedRolePolicies)
		r.Get("/roles/{roleName}/policies/{policyName}", h.getRolePolicy)
		r.Get("/roles/{roleName}/inline-policies", h.listRolePolicies)

		r.Get("/policies", h.listPolicies)
		r.Post("/policies", h.createPolicy)
		r.Post("/policies/get", h.getPolicy)
		r.Post("/policies/delete", h.deletePolicy)

		r.Post("/groups", h.createGroup)
		r.Get("/groups", h.listGroups)
		r.Get("/groups/{groupName}", h.getGroup)
		r.Delete("/groups/{groupName}", h.deleteGroup)
		r.Post("/groups/{groupName}/users", h.addUserToGroup)
		r.Delete("/groups/{groupName}/users/{userName}", h.removeUserFromGroup)
		r.Get("/groups/{groupName}/users", h.listUsersForGroup)
		r.Get("/users/{userName}/groups", h.listGroupsForUser)
		r.Get("/users/{userName}/policies", h.listUserPolicies)
	})
}

func (h *ProxyHandler) createUser(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	input := &iam.CreateUserInput{}
	if err := parseBody(bodyBytes, input); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	result, err := h.Svc.IAM().CreateUser(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to create user", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) getUser(w http.ResponseWriter, r *http.Request) {
	input := &iam.GetUserInput{
		UserName: aws.String(chi.URLParam(r, "userName")),
	}
	result, err := h.Svc.IAM().GetUser(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to get user", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) listUsers(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	input := &iam.ListUsersInput{}
	if err := parseBody(bodyBytes, input); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	result, err := h.Svc.IAM().ListUsers(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to list users", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) deleteUser(w http.ResponseWriter, r *http.Request) {
	input := &iam.DeleteUserInput{
		UserName: aws.String(chi.URLParam(r, "userName")),
	}
	result, err := h.Svc.IAM().DeleteUser(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to delete user", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) createRole(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	input := &iam.CreateRoleInput{}
	if err := parseBody(bodyBytes, input); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	result, err := h.Svc.IAM().CreateRole(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to create role", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) getRole(w http.ResponseWriter, r *http.Request) {
	input := &iam.GetRoleInput{
		RoleName: aws.String(chi.URLParam(r, "roleName")),
	}
	result, err := h.Svc.IAM().GetRole(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to get role", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) listRoles(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	input := &iam.ListRolesInput{}
	if err := parseBody(bodyBytes, input); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	result, err := h.Svc.IAM().ListRoles(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to list roles", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) deleteRole(w http.ResponseWriter, r *http.Request) {
	input := &iam.DeleteRoleInput{
		RoleName: aws.String(chi.URLParam(r, "roleName")),
	}
	result, err := h.Svc.IAM().DeleteRole(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to delete role", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) listPolicies(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	input := &iam.ListPoliciesInput{}
	if err := parseBody(bodyBytes, input); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	result, err := h.Svc.IAM().ListPolicies(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to list policies", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) getPolicy(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	var body struct {
		PolicyArn string `json:"PolicyArn"`
	}
	if err := parseBody(bodyBytes, &body); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	input := &iam.GetPolicyInput{
		PolicyArn: aws.String(body.PolicyArn),
	}
	result, err := h.Svc.IAM().GetPolicy(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to get policy", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) createPolicy(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	input := &iam.CreatePolicyInput{}
	if err := parseBody(bodyBytes, input); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	result, err := h.Svc.IAM().CreatePolicy(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to create policy", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) deletePolicy(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	var body struct {
		PolicyArn string `json:"PolicyArn"`
	}
	if err := parseBody(bodyBytes, &body); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	input := &iam.DeletePolicyInput{
		PolicyArn: aws.String(body.PolicyArn),
	}
	result, err := h.Svc.IAM().DeletePolicy(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to delete policy", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) createAccessKey(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	input := &iam.CreateAccessKeyInput{}
	if err := parseBody(bodyBytes, input); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	result, err := h.Svc.IAM().CreateAccessKey(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to create access key", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) listAccessKeys(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	input := &iam.ListAccessKeysInput{}
	if err := parseBody(bodyBytes, input); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	result, err := h.Svc.IAM().ListAccessKeys(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to list access keys", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) deleteAccessKey(w http.ResponseWriter, r *http.Request) {
	input := &iam.DeleteAccessKeyInput{
		AccessKeyId: aws.String(chi.URLParam(r, "accessKeyId")),
	}
	result, err := h.Svc.IAM().DeleteAccessKey(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to delete access key", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) updateAccessKeyStatus(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	var body struct {
		Status string `json:"Status"`
	}
	if err := parseBody(bodyBytes, &body); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	input := &iam.UpdateAccessKeyInput{
		AccessKeyId: aws.String(chi.URLParam(r, "accessKeyId")),
		Status:      iamtypes.StatusType(body.Status),
	}
	result, err := h.Svc.IAM().UpdateAccessKeyStatus(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to update access key status", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) attachRolePolicy(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	var body struct {
		PolicyArn string `json:"PolicyArn"`
	}
	if err := parseBody(bodyBytes, &body); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	input := &iam.AttachRolePolicyInput{
		RoleName:  aws.String(chi.URLParam(r, "roleName")),
		PolicyArn: aws.String(body.PolicyArn),
	}
	result, err := h.Svc.IAM().AttachRolePolicy(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to attach role policy", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) detachRolePolicy(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	var body struct {
		PolicyArn string `json:"PolicyArn"`
	}
	if err := parseBody(bodyBytes, &body); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	input := &iam.DetachRolePolicyInput{
		RoleName:  aws.String(chi.URLParam(r, "roleName")),
		PolicyArn: aws.String(body.PolicyArn),
	}
	result, err := h.Svc.IAM().DetachRolePolicy(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to detach role policy", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) listAttachedRolePolicies(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	var body struct {
		PathPrefix string `json:"PathPrefix"`
		Marker     string `json:"Marker"`
		MaxItems   int32  `json:"MaxItems"`
	}
	if err := parseBody(bodyBytes, &body); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	input := &iam.ListAttachedRolePoliciesInput{
		RoleName: aws.String(chi.URLParam(r, "roleName")),
	}
	if body.PathPrefix != "" {
		input.PathPrefix = aws.String(body.PathPrefix)
	}
	if body.Marker != "" {
		input.Marker = aws.String(body.Marker)
	}
	if body.MaxItems > 0 {
		input.MaxItems = aws.Int32(body.MaxItems)
	}
	result, err := h.Svc.IAM().ListAttachedRolePolicies(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to list attached role policies", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) createGroup(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	input := &iam.CreateGroupInput{}
	if err := parseBody(bodyBytes, input); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	result, err := h.Svc.IAM().CreateGroup(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to create group", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) getGroup(w http.ResponseWriter, r *http.Request) {
	input := &iam.GetGroupInput{
		GroupName: aws.String(chi.URLParam(r, "groupName")),
	}
	result, err := h.Svc.IAM().GetGroup(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to get group", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) listGroups(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	input := &iam.ListGroupsInput{}
	if err := parseBody(bodyBytes, input); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	result, err := h.Svc.IAM().ListGroups(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to list groups", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) deleteGroup(w http.ResponseWriter, r *http.Request) {
	input := &iam.DeleteGroupInput{
		GroupName: aws.String(chi.URLParam(r, "groupName")),
	}
	result, err := h.Svc.IAM().DeleteGroup(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to delete group", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) addUserToGroup(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	var body struct {
		UserName string `json:"UserName"`
	}
	if err := parseBody(bodyBytes, &body); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	input := &iam.AddUserToGroupInput{
		GroupName: aws.String(chi.URLParam(r, "groupName")),
		UserName:  aws.String(body.UserName),
	}
	result, err := h.Svc.IAM().AddUserToGroup(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to add user to group", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) removeUserFromGroup(w http.ResponseWriter, r *http.Request) {
	input := &iam.RemoveUserFromGroupInput{
		GroupName: aws.String(chi.URLParam(r, "groupName")),
		UserName:  aws.String(chi.URLParam(r, "userName")),
	}
	result, err := h.Svc.IAM().RemoveUserFromGroup(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to remove user from group", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) listGroupsForUser(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	var body struct {
		Marker   string `json:"Marker"`
		MaxItems int32  `json:"MaxItems"`
	}
	if err := parseBody(bodyBytes, &body); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	input := &iam.ListGroupsForUserInput{
		UserName: aws.String(chi.URLParam(r, "userName")),
	}
	if body.Marker != "" {
		input.Marker = aws.String(body.Marker)
	}
	if body.MaxItems > 0 {
		input.MaxItems = aws.Int32(body.MaxItems)
	}
	result, err := h.Svc.IAM().ListGroupsForUser(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to list groups for user", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) listUsersForGroup(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	var body struct {
		Marker   string `json:"Marker"`
		MaxItems int32  `json:"MaxItems"`
	}
	if err := parseBody(bodyBytes, &body); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	input := &iam.GetGroupInput{
		GroupName: aws.String(chi.URLParam(r, "groupName")),
	}
	if body.Marker != "" {
		input.Marker = aws.String(body.Marker)
	}
	if body.MaxItems > 0 {
		input.MaxItems = aws.Int32(body.MaxItems)
	}
	result, err := h.Svc.IAM().GetGroup(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to get group", err)
		return
	}
	writeJSON(w, http.StatusOK, map[string]interface{}{
		"Users":       result.Users,
		"IsTruncated": result.IsTruncated,
		"Marker":      result.Marker,
	})
}

func (h *ProxyHandler) listUserPolicies(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	var body struct {
		Marker   string `json:"Marker"`
		MaxItems int32  `json:"MaxItems"`
	}
	if err := parseBody(bodyBytes, &body); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	input := &iam.ListUserPoliciesInput{
		UserName: aws.String(chi.URLParam(r, "userName")),
	}
	if body.Marker != "" {
		input.Marker = aws.String(body.Marker)
	}
	if body.MaxItems > 0 {
		input.MaxItems = aws.Int32(body.MaxItems)
	}
	result, err := h.Svc.IAM().ListUserPolicies(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to list user policies", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) listRolePolicies(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	var body struct {
		Marker   string `json:"Marker"`
		MaxItems int32  `json:"MaxItems"`
	}
	if err := parseBody(bodyBytes, &body); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	input := &iam.ListRolePoliciesInput{
		RoleName: aws.String(chi.URLParam(r, "roleName")),
	}
	if body.Marker != "" {
		input.Marker = aws.String(body.Marker)
	}
	if body.MaxItems > 0 {
		input.MaxItems = aws.Int32(body.MaxItems)
	}
	result, err := h.Svc.IAM().ListRolePolicies(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to list role policies", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) getRolePolicy(w http.ResponseWriter, r *http.Request) {
	input := &iam.GetRolePolicyInput{
		RoleName:   aws.String(chi.URLParam(r, "roleName")),
		PolicyName: aws.String(chi.URLParam(r, "policyName")),
	}
	result, err := h.Svc.IAM().GetRolePolicy(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to get role policy", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}
